File: src/utils/taxTest.test.ts
import { describe, expect, it } from 'vitest'
import cases from '../../test-fixtures/tax/cases.json'
import { lotTaxStatus, realizedByYear, timeTestDate } from './taxTest'
import type { Position } from '../types'

describe('Czech time test (shared cases)', () => {
  it.each(cases.freeFrom)('free from %j', ({ buyDate, expected }) => {
    expect(timeTestDate(buyDate)).toBe(expected)
  })

  it.each(cases.status)('status %j', ({ lot, today, kind, daysLeft }) => {
    const s = lotTaxStatus(lot, today)
    expect(s.kind).toBe(kind)
    if (daysLeft !== undefined) expect(s.daysLeft).toBe(daysLeft)
  })

  it('sums realized gains per year in CZK at trade-date rates', () => {
    const fx = cases.years.fx as unknown as Record<string, [string, number][]>
    const czkAt = (amount: number, cur: string, date: string) => {
      if (cur === 'CZK') return amount
      const rate = [...fx[cur]].reverse().find(([d]) => d <= date)![1]
      return amount * rate
    }
    const got = realizedByYear(cases.years.positions as Position[], czkAt)
    expect(got).toEqual(cases.years.expected)
  })
})

File: android/core/calc/src/test/kotlin/com/stocktracker/core/calc/TaxTestTest.kt
package com.stocktracker.core.calc

import com.stocktracker.core.model.Position
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.boolean
import kotlinx.serialization.json.double
import kotlinx.serialization.json.int
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.long
import org.junit.Assert.assertEquals
import org.junit.Test
import java.io.File

/** Runs the same case table as src/utils/taxTest.test.ts. */
class TaxTestTest {
    private val json = Json { ignoreUnknownKeys = true }
    private val cases = json.parseToJsonElement(
        File(System.getProperty("user.dir")).resolve("../../../test-fixtures/tax/cases.json").readText(),
    ).jsonObject

    @Test fun freeFrom() = cases.getValue("freeFrom").jsonArray.forEach {
        val c = it.jsonObject
        assertEquals(c.getValue("expected").jsonPrimitive.content, timeTestDate(c.getValue("buyDate").jsonPrimitive.content))
    }

    @Test fun status() = cases.getValue("status").jsonArray.forEach {
        val c = it.jsonObject
        val lot = c.getValue("lot").jsonObject
        val s = lotTaxStatus(
            buyDate = lot.getValue("buyDate").jsonPrimitive.content,
            sellDate = lot["sellDate"]?.jsonPrimitive?.content,
            sellPrice = lot["sellPrice"]?.jsonPrimitive?.double,
            today = c.getValue("today").jsonPrimitive.content,
        )
        assertEquals(c.toString(), c.getValue("kind").jsonPrimitive.content.uppercase(), s.kind.name)
        c["daysLeft"]?.let { d -> assertEquals(c.toString(), d.jsonPrimitive.long, s.daysLeft) }
    }

    @Test fun years() {
        val y = cases.getValue("years").jsonObject
        val fx = y.getValue("fx").jsonObject.mapValues { (_, v) ->
            v.jsonArray.map { p -> p.jsonArray[0].jsonPrimitive.content to p.jsonArray[1].jsonPrimitive.double }
        }
        val czkAt = { amount: Double, cur: String, date: String ->
            if (cur == "CZK") amount else amount * fx.getValue(cur).last { it.first <= date }.second
        }
        val positions = json.decodeFromJsonElement(kotlinx.serialization.builtins.ListSerializer(Position.serializer()), y.getValue("positions"))
        val got = realizedByYear(positions, czkAt)
        val expected = y.getValue("expected").jsonArray.map { e ->
            val o = e.jsonObject
            YearSummary(
                o.getValue("year").jsonPrimitive.int, o.getValue("proceedsTaxable").jsonPrimitive.double,
                o.getValue("gainTaxable").jsonPrimitive.double, o.getValue("gainExempt").jsonPrimitive.double,
                o.getValue("underSmallProceedsLimit").jsonPrimitive.boolean,
            )
        }
        assertEquals(expected, got)
    }
}
